### PR TITLE
Modify qchem functions to separately take the atomic symbols and coordinates

### PR DESCRIPTION
**Pull request by agran2018**
**Context:**
In the ADR ["Add functionality for computing energy gradients"](https://github.com/PennyLaneAI/adrs/blob/energy-gradients/documents/042-energy-gradients.rst) we proposed to update the ``molecular_hamiltonian()`` function to separately take as input a list of strings representing the atomic symbols and an array containing the atomic positions in Cartesian coordinates.

```python
import pennylane as qml
symbols, coordinates = (['H', 'H'], np.array([0., 0., -0.661, 0., 0., 0.661]))
H, qubits = qml.qchem.molecular_hamiltonian(symbols, coordinates)
```

**Description of the Change:**
-  Modify ``read_structure()`` function to output a list of atomic symbols and a 1D array with the positions of the atoms in _atomic units_.
-  Modify ``meanfield()`` and ``molecular_hamiltonian()`` functions to handle the new convention to store the molecular geometry.
-  Make ``name`` a keyword argument in ``meanfield`` and ``molecular_hamiltonian`` functions.
- Update unit tests.

**Benefits:**
The user can create a callable function to evaluate parametrized electronic Hamiltonians.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: qchem/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,12 @@
 
 <h3>Improvements</h3>
 
+* Updates `PennyLane-QChem` functions to separately take as input the
```
Comment (josh146) [inline qchem/CHANGELOG.md:7]: Suggest removing mention of `PennyLane-QChem`, and instead listing the functions that have changed with bullet points :)

File: qchem/pennylane_qchem/qchem/structure.py [hunk 2]
```diff
@@ -176,12 +177,14 @@ def read_structure(filepath, outpath="."):
...
-        list: for each atomic species, a list containing the symbol and the Cartesian coordinates
+        tuple[list, array: symbols of the atoms in the molecule and their positions in Cartesian
+        coordinates.
 
     **Example**
 
-    >>> read_structure('h2_ref.xyz')
-    [['H', (0.0, 0.0, -0.35)], ['H', (0.0, 0.0, 0.35)]]
+    >>> symbols, coordinates = read_structure('h2.xyz')
+    >>> print(symbols coordinates)
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/structure.py:186]:
```suggestion
    >>> print(symbols, coordinates)
```

File: qchem/pennylane_qchem/qchem/structure.py [hunk 3]
```diff
@@ -176,12 +177,14 @@ def read_structure(filepath, outpath="."):
         outpath (str): path to the output directory
 
     Returns:
-        list: for each atomic species, a list containing the symbol and the Cartesian coordinates
+        tuple[list, array: symbols of the atoms in the molecule and their positions in Cartesian
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/structure.py:180]:
```suggestion
        tuple[list, array]: symbols of the atoms in the molecule and their positions in Cartesian
```

File: qchem/CHANGELOG.md (hunk 1 above)
Comment (josh146) [reply qchem/CHANGELOG.md:7]: The changelog is a good place to provide all the fine-grained detail. Imagine a user reading *just* the changelog - you'll want them to be able to walk away knowing what changes they'll need to make to their code.

File: qchem/pennylane_qchem/qchem/structure.py [hunk 4]
```diff
@@ -248,7 +256,10 @@ def meanfield(
 
     Args:
         name (str): molecule label
-        geometry (list): list containing the symbol and Cartesian coordinates for each atom
+        symbols (list[str]): symbols of the atomic species in the molecule
+        coordinates (array[float]): 1D array with the atomic positions in Cartesian
+            coordinates. The coordinates must be given in Angstroms and the size of the array
+            should be ``3*N`` with ``N`` being the number of atoms.
```
Comment (ixfoduap) [inline qchem/pennylane_qchem/qchem/structure.py:262]:
```suggestion
            should be ``3*N`` where ``N`` is the number of atoms.
```

File: qchem/pennylane_qchem/qchem/structure.py [hunk 5]
```diff
@@ -267,11 +278,17 @@ def meanfield(
...
 
     >>> name = 'h2'
-    >>> geometry = [['H', (0.0, 0.0, -0.35)], ['H', (0.0, 0.0, 0.35)]]
-    >>> meanfield(name, geometry)
+    >>> symbols, coordinates = (['H', 'H'], np.array([ 0., 0., -0.35, 0., 0., 0.35]))
+    >>> meanfield(name, symbols, coordinates)
     ./pyscf/sto-3g/h2
     """
 
+    if coordinates.size != 3 * len(symbols):
```
Comment (ixfoduap) [inline qchem/pennylane_qchem/qchem/structure.py:286]: Great check!

File: qchem/pennylane_qchem/qchem/structure.py [hunk 6]
```diff
@@ -728,7 +744,10 @@ def molecular_hamiltonian(
 
     Args:
         name (str): name of the molecule
-        geo_file (str): file containing the geometry of the molecule
+        symbols (list[str]): symbols of the atomic species in the molecule
+        coordinates (array[float]): 1D array with the atomic positions in Cartesian
+            coordinates. The coordinates must be given in Angstroms and the size of the array
+            should be ``3*N`` with ``N`` being the number of atoms.
```
Comment (ixfoduap) [inline qchem/pennylane_qchem/qchem/structure.py:750]:
```suggestion
            should be ``3*N`` where ``N`` is the number of atoms.
```

File: qchem/pennylane_qchem/qchem/structure.py [hunk 7]
```diff
@@ -248,7 +256,10 @@ def meanfield(
 
     Args:
         name (str): molecule label
```
Comment (ixfoduap) [inline qchem/pennylane_qchem/qchem/structure.py:258]: I'm still confused about having to pass `name` as an argument. In my mind, it should be possible for this function to work without having to create any local files. Am I wrong? Is this an issue related to Psi4 and PySCF?

**Review by ixfoduap — COMMENTED** [review]
Looks great! Just left a few minor suggestions and a question related the `name` argument in `molecular_hamiltonian`

File: qchem/pennylane_qchem/qchem/structure.py (hunk 7 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/structure.py:258]: PySCF/Psi4 creates an HDF5-format file from which the HF quantities required to build H are retrieved. We offer the user the possibility to label the molecule in order not to overwrite this file for different molecules. 
For example, if you want to perform quantum simulations of different approximations to the Hamiltonian (e.g., different active orbitals) of a given molecule you want to make sure you are reading the right mean field input:

```
hf_file = "./pyscf/6-31g/h2o.hdf5"
H = qchem.convert_observable(qchem.decompose(hf_file, mapping="jordan_wigner", core=... , active=...))
```  
so you don't need to run the HF again. If we think this feature is mainly useful for advanced users we could make ``name`` a keyword argument in ``meanfield()`` and ``molecular_hamiltonian()``.

```
def meanfield(symbols, coordinates, name="mol", charge=0, mult=1, basis="sto-3g", package="pyscf", outpath=".") 
```
what do you think?. I like it.

Comment (ixfoduap) [reply qchem/pennylane_qchem/qchem/structure.py:258]: I like it too! 

But to be clear, Psi4 and PySCF always create these local files? This is not something that can be avoided?

Comment (agran2018) [reply qchem/pennylane_qchem/qchem/structure.py:258]: I don't think so because OpenFermion uses this file to load the quantities we need to build Hamiltonian.

Comment (josh146) [reply qchem/pennylane_qchem/qchem/structure.py:258]: Is there a way to 'skip' this, though? OpenFermion does not allow you to build a Hamiltonian using Python objects outputted by Psi4/pyscf?

Comment (agran2018) [reply qchem/pennylane_qchem/qchem/structure.py:258]: OpenFermion-Psi4/Pyscf plugins add python scripts to the input files of Psi4/pyscf to fish the electronic integrals and some other things and output them to the HDF5-format file which is used to populate the attributes of the ``MolecularData`` class (https://github.com/quantumlib/OpenFermion/blob/master/src/openfermion/chem/molecular_data.py)

File: qchem/tests/test_molecular_hamiltonian.py [hunk 8]
```diff
@@ -6,20 +6,47 @@
 
 from pennylane.vqe import Hamiltonian
 
-ref_dir = os.path.join(os.path.dirname(os.path.realpath(__file__)), "test_ref_files")
+import numpy as np
+
+
+symbols = ["C", "C", "N", "H", "H", "H", "H", "H"]
+coordinates = np.array(
```
Comment (soranjh) [inline qchem/tests/test_molecular_hamiltonian.py:13]: Can this be given in one line?

File: qchem/tests/test_read_structure.py [hunk 9]
```diff
@@ -32,21 +33,41 @@ def test_conversion_from_folder(names, tmpdir, requires_babel):
...
-        ["C", (-0.714, 0.125, 0.327)],
-        ["N", (0.683, 0.133, 0.745)],
-        ["H", (0.442, -1.529, -0.619)],
-        ["H", (0.672, 0.102, -1.428)],
-        ["H", (-1.364, -0.56, 0.857)],
-        ["H", (-1.149, 1.08, 0.060)],
-        ["H", (1.093, 1.063, 0.636)],
-    ]
+    ref_symbols = ["C", "C", "N", "H", "H", "H", "H", "H"]
+    ref_coords = np.array(
```
Comment (soranjh) [inline qchem/tests/test_read_structure.py:37]: Can this be given in one line?

File: qchem/CHANGELOG.md [hunk 10]
```diff
@@ -4,6 +4,28 @@
...
+  a 1D array with the atomic positions in Cartesian coordinates.
+
+  - The ``read_structure`` function returns a list with the symbols of the atoms and
+    the array with the atomic positions. 
+
+  - The ``meanfield`` and ``molecular_hamiltonian`` functions takes separately the
+    list of atomic symbols and the array with the atomic coordinates.
+
+  - Labeling the molecule is now optional as we have made ``name`` a keyword argument
+    in ``meanfield`` and ``molecular_hamiltonian`` functions.
```
Comment (ixfoduap) [inline qchem/CHANGELOG.md:17]:
```suggestion
  - Labelling the molecule is now optional as we have made ``name`` a keyword argument
    in the ``meanfield`` and ``molecular_hamiltonian`` functions.
```

File: qchem/CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,28 @@
...
+
+    For example:
+
+    ```pycon
+    >>> symbols, coordinates = (['H', 'H'], np.array([ 0., 0., -0.35, 0., 0., 0.35]))
+    >>> H, qubits = qml.qchem.molecular_hamiltonian(symbols, coordinates)
+    ```
+  This allows users to build parametrized electronic Hamiltonians to perform 
+  optimization of the energy with respect to the Hamiltonian parameters.
+  [(#1078)](https://github.com/PennyLaneAI/pennylane/pull/1078)
```
Comment (ixfoduap) [inline qchem/CHANGELOG.md:27]:
```suggestion
  This allows users to more easily build parametrized electronic Hamiltonians
  [(#1078)](https://github.com/PennyLaneAI/pennylane/pull/1078)
```

**Review by ixfoduap — APPROVED** [review]

File: qchem/tests/test_molecular_hamiltonian.py (hunk 8 above)
Comment (agran2018) [reply qchem/tests/test_molecular_hamiltonian.py:13]: I wrote it in one line by "black" changed it to a column :-)

**Review by soranjh — APPROVED** [review]
Looks good to me. Left few comments.

File: qchem/tests/test_read_structure.py (hunk 9 above)
Comment (agran2018) [reply qchem/tests/test_read_structure.py:37]: I wrote it in one line by "black" changed it to a column :-)

File: qchem/CHANGELOG.md [hunk 12]
```diff
@@ -4,6 +4,27 @@
...
+    in the ``meanfield`` and ``molecular_hamiltonian`` functions.
+
+    For example:
+
+    ```pycon
+    >>> symbols, coordinates = (['H', 'H'], np.array([ 0., 0., -0.35, 0., 0., 0.35]))
+    >>> H, qubits = qml.qchem.molecular_hamiltonian(symbols, coordinates)
+    ```
+  This allows users to more easily build parametrized electronic Hamiltonians
+  [(#1078)](https://github.com/PennyLaneAI/pennylane/pull/1078)
```
Comment (josh146) [inline qchem/CHANGELOG.md:26]: much better! 💯

**Comment by zeyueN** [comment]
Just saw this on slack and only read the description. It looks like the interface change intends to save some state (`symbols` etc.) for the of users. There's perhaps already a reason somewhere in the ADR, but I'm just curious why `partial` is used instead of making `H` a `class` instance?

i.e.
```
class MolecularHamiltonian:
  def __init__(self, symbols, basis):
    self.symbols=symbols
    self.basis=basis

  def __call__(self, x):
    return qml.qchem.molecular_hamiltonian(self.symbols, x, basis=self.basis)[0]

H = MolecularHamiltonian(symbols, basis)
```

**Comment by agran2018** [comment]
> Just saw this on slack and only read the description. It looks like the interface change intends to save some state (`symbols` etc.) for the of users. There's perhaps already a reason somewhere in the ADR, but I'm just curious why `partial` is used instead of making `H` a `class` instance?
> 
> i.e.
> 
> ```
> class MolecularHamiltonian:
>   def __init__(self, symbols, basis):
>     self.symbols=symbols
>     self.basis=basis
> 
>   def __call__(self, x):
>     return qml.qchem.molecular_hamiltonian(self.symbols, x, basis=self.basis)[0]
> 
> H = MolecularHamiltonian(symbols, basis)
> print(H(x))
> ```

Thank you @zeyueN for your sugestion and the simple example that helps me to understand how a class looks like in python :-). Just a comment, it is not mandatory to use partial, one can also do:

```
symbols = ["H", "H"]

def H(x):
    return qml.qchem.molecular_hamiltonian(symbols, x, basis="sto-6g", mapping="bravyi_kitaev")[0]
```
Apart from the fact that having a class looks cleaner, what other advantage does it bring?

**Comment by zeyueN** [comment]
The general answer to that question is that it's more "pythonic" 😁 . But really it's just a design decision among many technically equivalent implementations. When I see the use of `partial`, I see a need to save some state, and class is python's natural way to do it. In my simple class, there's not too much advantage yet, other than the ability to get/set `H.symbols` and `H.basis`, which you may or may not want. You can of course extend the class with more methods that are natural to call from an H object to make it more powerful.

My question really came from the user experience perspective. Most python users are used to playing with objects, in ipython or jupyter, where you can use tab completion to quickly see what associated properties/methods are available within an object. And a lot of times I find it's easier to think about calculations to be associated with objects (e.g. say, I don't know, molecule_0.find_ground_state(), ansatz_2.optimize()). But again it really depends on what the users (quantum chemists) are used to and what conventions they like to follow.

File: qchem/CHANGELOG.md (hunk 12 above)
Comment (josh146) [inline qchem/CHANGELOG.md:26]:
```suggestion

  This allows users to more easily build parametrized electronic Hamiltonians
  [(#1078)](https://github.com/PennyLaneAI/pennylane/pull/1078)
```

**Comment by josh146** [comment]
Hi @agran2018 -- I think @zeyueN raises a very good point.

Just to understand a bit better, how come you use `partial` in the PR example? Is there a reason for ensuring that `H` only has one argument `x`? Put another way, why not write it as follows?

```python
import pennylane as qml
from pennylane import numpy as np

# define H(x)
symbols, x = (['H', 'H'], np.array([ 0., 0., -0.35, 0., 0., 0.35]))
print(qml.qchem.molecular_hamiltonian(symbols, x, basis="6-31g"))
(2.2947063590000547) [I0]
+ (-0.2714597809530114) [Z0]
+ (0.08861654589379706) [Y0 Z1 Z2 Z3 Y4]
```

**Comment by agran2018** [comment]
> Hi @agran2018 -- I think @zeyueN raises a very good point.
> 
> Just to understand a bit better, how come you use `partial` in the PR example? Is there a reason for ensuring that `H` only has one argument `x`? Put another way, why not write it as follows?
> 
> ```
> >>> import pennylane as qml
> >>> from pennylane import numpy as np
> >>> symbols, x = (['H', 'H'], np.array([ 0., 0., -0.35, 0., 0., 0.35]))
> >>> qml.qchem.molecular_hamiltonian(symbols, x, basis="6-31g")
>   (2.2947063590000547) [I0]
>   + (-0.2714597809530114) [Z0]
>   + (0.08861654589379706) [Y0 Z1 Z2 Z3 Y4]
> ```

Yes, this is also perfectly fine. Indeed, for the algorithms we have been prototyping there is no need of using the `` partial`` function. The purpose in the PR example was to illustrate that one of the motivation to decouple the atomic symbols and the nuclear coordinates is allow users building parametrized electronic Hamiltonian ``H(x)``.

File: qchem/pennylane_qchem/qchem/structure.py [hunk 13]
```diff
@@ -148,14 +152,15 @@ def _exec_exists(prog):
 def read_structure(filepath, outpath="."):
     r"""Reads the structure of the polyatomic system from a file and returns
     a list with the symbols of the atoms in the molecule and a 1D array
-    with the atomic positions in Cartesian coordinates.
+    with their positions :math:`[x_1, y_1, z_1, x_2, y_2, z_2, \dots]` in
+    atomic units (Bohr).
```
Comment (ixfoduap) [inline qchem/pennylane_qchem/qchem/structure.py:156]:
```suggestion
    atomic units (Bohr radius = 1).
```

File: qchem/CHANGELOG.md [hunk 14]
```diff
@@ -4,6 +4,29 @@
 
 <h3>Improvements</h3>
 
+* The molecular geometry is now stored by a list containing the atomic symbols and
+  a 1D array with the position of the atoms in atomic units.
+
+  - The ``read_structure`` function returns a list with the symbols of the atoms and
+    the array with the atomic positions. The size of the array should be ``3*N`` where
+    ``N`` is the number of atoms.
```
Comment (ixfoduap) [inline qchem/CHANGELOG.md:12]:
```suggestion
    the array with the atomic positions. 
```

**Review by ixfoduap — APPROVED** [review]

**Review by soranjh — APPROVED** [review]

**Review by josh146 — APPROVED** [review]
